### ISSUE TITLE
[CI] ReadinessClusterIT testReadinessDuringRestarts failing

### ISSUE THREAD

**Issue by mark-vieira** [elastic/elasticsearch#91434]
Did not reproduce.

**Build scan:**
https://gradle-enterprise.elastic.co/s/36ijsjujdw2xe/tests/:server:internalClusterTest/org.elasticsearch.readiness.ReadinessClusterIT/testReadinessDuringRestarts

**Reproduction line:**
`./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.readiness.ReadinessClusterIT.testReadinessDuringRestarts" -Dtests.seed=5BB624F761E4E09B -Dtests.locale=ar-SY -Dtests.timezone=JST -Druntime.java=17`

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.readiness.ReadinessClusterIT&tests.test=testReadinessDuringRestarts

**Failure excerpt:**
```
junit.framework.AssertionFailedError: Expected exception IOException but no exception was thrown

  at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2854)
  at org.apache.lucene.tests.util.LuceneTestCase.expectThrows(LuceneTestCase.java:2840)
  at org.elasticsearch.test.readiness.ReadinessClientProbe.lambda$tcpReadinessProbeFalse$2(ReadinessClientProbe.java:73)
  at java.security.AccessController.doPrivileged(AccessController.java:318)
  at org.elasticsearch.test.readiness.ReadinessClientProbe.tcpReadinessProbeFalse(ReadinessClientProbe.java:72)
  at org.elasticsearch.readiness.ReadinessClusterIT.testReadinessDuringRestarts(ReadinessClusterIT.java:86)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

**Comment by grcevski** [comment]
My current theory here is that because we use the ephemeral port 0 for testing, so that we don't experience port bind clashes, a concurrent test reused the port (after we stopped the master) and it was open when we attempted to connect.

**Comment by pgomulka** [comment]
> We don't use port 0 for internal cluster tests. We use a distinct port range that's designed to minimize overlap as much as possible. It's still possible for their to be conflicts and reuse but pretty rare. See https://github.com/elastic/elasticsearch/pull/87885.

But you are referring to transport and http ports.

In this Internal cluster test the problem is with the readiness probe port which is configured 0.
https://github.com/elastic/elasticsearch/blob/b86ec3e86a3e2255703a3118083dc6e331605e0c/server/src/internalClusterTest/java/org/elasticsearch/readiness/ReadinessClusterIT.java#L85

Similarly it is using port 0 in test clusters
https://github.com/elastic/elasticsearch/blob/f21881369c849011e6aa2621d64627e4e0669d69/build-tools/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchCluster.java#L152


But agree that we should try to use the same mechanism with port ranges once we confirm/reproduce this problem.

**Comment by rjernst** [comment]
I think abstracting the socket access is better for testability. Note that while we stopped using mock nio transport for InternalClusterTest, we still have mock transports for some tests. Conversely, is there any reason a port range is useful outside of tests? I can't think of any, in which case having a port range seems like test code bleeding into production.

**Comment by ldematte** [comment]
I have spent some time today investigating this/related issues/potential problems with the metering service tests (which also use an ephemeral port as pointed out by @pgomulka).
Even if initially I thought that the second solution proposed by @pgomulka was better for uniformity (it is how the problem is tackled elsewhere after all), after some digging I have to agree with @rjernst: for this service, the port range does not make sense for the real use, it will be confusing, and will be used only in tests. 

It's very likely that this "ephemeral ports in tests" is one of those problems where there is no "one fits all" solution, needs to be solved case by case, so I concentrated on this.

The test is covering the quite complex lifecycle of the `ServerSocketChannel` and associated `SocketAddress` during the service readiness cycle.

The `ServerSocketChannel` gets created and destroyed at every start/stop, but the `SocketAddress` is created/set once and reused (to bind to the same port over and over again). 
We could inject a `ServerSocketChannel` factory (Supplier) which returns a (mocked) `ServerSocketChannel` paired with a (mocked) `ServerChannel` that will keep the mocked socket status.
We could even skip mocking the `ServerChannel` - that is used just to test the socket status by probing the tcp port by a couple of functions (`tcpReadinessProbeTrue` and `tcpReadinessProbeTrue`), and we could replace them with direct checks on the object holding mocked socket status.

I realized after writing that probably this is exactly what @pgomulka wanted to do with 1. 
If you both are OK with that, I could give it a try!

**Comment by rjernst** [comment]
@ldematte Your plan sounds good to me, I would just strive to mock the already existing jdk classes rather than adding any additional abstraction.

**Comment by ldematte** [comment]
Some notes on how it went so far:
We need to mock a static method, `ServerSocketChannel::open`.
As a first step I tried Mockito, I used it before to mock static methods. Unfortunately, we are using a `SubclassMockMaker`, which does not allow mock of static and final. We have a wrapper for in (`SecureMockMaker`) and there is even a TODO comment from Ryan that we should consider the `Inline` MockMaker. That will allow us to mock final and static, but (as the comment states too) this requires a Java agent.

Mockito/ByteBuddy to provide a Java agent that _should_ do things automatically; however it requires starting an external process, which fails due to the security manager.

`SecureMockMaker` already does some work to make Mokito work with the SM; I have tried to modify the AccessControlContext it creates by cloning and adding the relevant permissions (file read/execute and manageProcess). This lead to the permissions check to pass, but the process launch still fails (with errno 2 (file not found) on macOS and errno 13 (permission denied) on Linux).

Since these are IT, which launch a real cluster (albeit internal), I suspect there is some other layer of protection involved (maybe cgroups or similar?): after all, launching processes is something we really want to limit.
I will investigate this further in the background.
Anyway, I was already undecided about this direction as switching MockMaker would have been a non-trivial change to the test framework; so I will try alternatives. Namely, try to insert an abstraction of opening a channel, defaulting to `ServerSocketChannel::open` but making it "pluggable" for tests. I was thinking a separate ctor (package private), but of course this is not an option since this is a IT test.

On Ryan suggestion I will look at how mock/test modules are loaded in `MockNode` and add a protected method + override in tests, let's see how it goes.
